### PR TITLE
Check if registration socket is still valid

### PR DESCRIPTION
**Pull request by gnufied**
Health check should check if registration socket is really responsive.

```release-note
Do not merely check presence of registration socket, but also check if it responds to grpc requests
```

### CODE REVIEW

File: cmd/csi-node-driver-registrar/node_register.go [hunk 1]
```diff
@@ -126,6 +136,43 @@ func httpServer(socketPath string, httpEndpoint string) {
...
+
+	klog.V(1).Infof("Calling node registrar to check if it still responds")
+	ctx, cancel := context.WithTimeout(context.Background(), *operationTimeout)
+	defer cancel()
+
+	client := registerapi.NewRegistrationClient(csiConn)
+
+	infoRequest := &registerapi.InfoRequest{}
+
+	info, err := client.GetInfo(ctx, infoRequest)
```
Comment (jsafrane) [inline cmd/csi-node-driver-registrar/node_register.go:157]: `Probe()` would be more appropriate

File: cmd/csi-node-driver-registrar/node_register.go [hunk 2]
```diff
@@ -126,6 +136,43 @@ func httpServer(socketPath string, httpEndpoint string) {
 	klog.Fatal(http.ListenAndServe(httpEndpoint, mux))
 }
 
+func checkLiveRegistrationSocket(socketFile, csiDriverName string) error {
+	klog.V(1).Infof("Attempting to open a gRPC connection with: %q", socketFile)
+	cmm := metrics.NewCSIMetricsManagerForSidecar("")
```
Comment (jsafrane) [inline cmd/csi-node-driver-registrar/node_register.go:141]: The manager should not be re-created for each health check. Please use the one from main.go

File: cmd/csi-node-driver-registrar/node_register.go [hunk 3]
```diff
@@ -126,6 +136,43 @@ func httpServer(socketPath string, httpEndpoint string) {
...
+	csiConn, err := connection.Connect(socketFile, cmm)
```
Comment (jsafrane) [inline cmd/csi-node-driver-registrar/node_register.go:142]: it's not CSI connection, it's registration socket

File: cmd/csi-node-driver-registrar/node_register.go [hunk 4]
```diff
@@ -126,6 +136,43 @@ func httpServer(socketPath string, httpEndpoint string) {
...
+	if err != nil {
+		return fmt.Errorf("error connecting to node-registrar socket %s: %v", socketFile, err)
+	}
+
+	defer closeGrpcConnection(socketFile, csiConn)
```
Comment (jsafrane) [inline cmd/csi-node-driver-registrar/node_register.go:147]: Does it make sense to close the connection? All CSI sidecars re-use the same connection for all requests. gRPC does some recovery when the socket is closed, AFAIK.

File: cmd/csi-node-driver-registrar/node_register.go [hunk 5]
```diff
@@ -126,6 +136,43 @@ func httpServer(socketPath string, httpEndpoint string) {
...
+
+	klog.V(1).Infof("Calling node registrar to check if it still responds")
```
Comment (jsafrane) [inline cmd/csi-node-driver-registrar/node_register.go:149]: V(2)

File: cmd/csi-node-driver-registrar/node_register.go [hunk 6]
```diff
@@ -126,6 +136,43 @@ func httpServer(socketPath string, httpEndpoint string) {
...
+
+	info, err := client.GetInfo(ctx, infoRequest)
+	if err != nil {
+		return fmt.Errorf("error getting info from node-registrar socket: %v", err)
+	}
+
+	if info.Name == csiDriverName {
+		return nil
+	}
+	return fmt.Errorf("invalid driver name %s", info.Name)
```
Comment (jsafrane) [inline cmd/csi-node-driver-registrar/node_register.go:165]: Will this be useful? Did we ever saw CSI drivers changing their name?

File: cmd/csi-node-driver-registrar/node_register.go (hunk 4 above)
Comment (gnufied) [reply cmd/csi-node-driver-registrar/node_register.go:147]: for sidecars I guess it makes sense. I am less sure about periodic checks. I think closing the connection here is fine.

File: cmd/csi-node-driver-registrar/node_register.go (hunk 6 above)
Comment (gnufied) [reply cmd/csi-node-driver-registrar/node_register.go:165]: I do not think driver name should change without requiring re-registration right?

File: cmd/csi-node-driver-registrar/node_register.go (hunk 2 above)
Comment (gnufied) [reply cmd/csi-node-driver-registrar/node_register.go]: I removed need for metrics manager. I think it was unncessary for this use case anyways.

File: cmd/csi-node-driver-registrar/node_register.go (hunk 3 above)
Comment (gnufied) [reply cmd/csi-node-driver-registrar/node_register.go]: renamed the socket file.

File: cmd/csi-node-driver-registrar/node_register.go [hunk 7]
```diff
@@ -126,6 +135,42 @@ func httpServer(socketPath string, httpEndpoint string) {
...
+func checkLiveRegistrationSocket(socketFile, csiDriverName string) error {
+	klog.V(2).Infof("Attempting to open a gRPC connection with: %q", socketFile)
+	grpcConn, err := connection.ConnectWithoutMetrics(socketFile)
+	if err != nil {
+		return fmt.Errorf("error connecting to node-registrar socket %s: %v", socketFile, err)
+	}
+
+	defer closeGrpcConnection(socketFile, grpcConn)
+
+	klog.V(1).Infof("Calling node registrar to check if it still responds")
```
Comment (jsafrane) [inline cmd/csi-node-driver-registrar/node_register.go:147]: nit: V(2)

**Comment by mowangdk** [comment]
We have a similar problem on our environment, csi-node-driver-registrar process still exists, but the socket has been deactivated, what is the root cause of the problem?  Is it an OS issue?

**Pull request merged**